horovod/torch/__init__.py

```diff
@@ -190,7 +204,7 @@ def broadcast_optimizer_state(optimizer, root_rank):
         for group in optimizer.param_groups:
             for p in group['params']:
                 p.grad = p.data.new(p.size()).zero_()
-        optimizer.step()
+        super(optimizer.__class__, optimizer).step()
```

Comment:
Any idea why does it fail with the original `optimizer.step()`?  Can you check the stack trace of the failure?

Reply:
It fails because of how ```test_broadcast_state()```  in test/torch.py works. At line 827, a new model and uninitialized optimizer are created. If the rank is 0, the previous optimizer state is loaded, then broadcast to other ranks. For ranks other than 0, the optimizer remains empty; therefore ```len(state_dict['state']) == 0``` , which causes it to call ```optimizer.step()``` in torch/__init__.py (the line I altered, line 207). In this case, the only purpose of calling ```optimizer.step()``` is to initialize an uninitialized optimizer. Since I force allreduce on all model parameters in ```_DistributedOptimizer().step()```, If ```optimizer.step()``` is not called on all ranks, resource starvation occurs.

Reply:
Ah, makes sense.  One concern I have is that some users may do this broadcast on the original optimizer before wrapping it in DistributedOptimizer.  Let's only do super if the optimizer is a subclass of _DistributedOptimizer?